pkg/controller/direct/sql/normalize.go

```diff
@@ -290,6 +303,95 @@ func normalizeAuditLogBucketRef(ctx context.Context, kube client.Reader, obj *kr
 	}
 }
 
+func normalizeSourceSQLInstanceRef(ctx context.Context, kube client.Reader, obj *krm.SQLInstance) (string, error) {
+	if obj.Spec.CloneSource == nil {
+		return "", nil
+	}
+
+	sqlInstanceRef := obj.Spec.CloneSource.SQLInstanceRef
+
+	if sqlInstanceRef.External != "" && sqlInstanceRef.Name != "" {
+		return "", fmt.Errorf("cannot specify both spec.settings.cloneSource.sqlInstanceRef.external and spec.settings.cloneSource.sqlInstanceRef.name")
+	}
+
+	if sqlInstanceRef.External != "" {
+		return sqlInstanceRef.External, nil
+	} else if sqlInstanceRef.Name != "" {
+		if sqlInstanceRef.Namespace == "" {
+			sqlInstanceRef.Namespace = obj.Namespace
+		}
+
+		key := types.NamespacedName{
+			Namespace: sqlInstanceRef.Namespace,
+			Name:      sqlInstanceRef.Name,
+		}
+
+		sqlInstance := &unstructured.Unstructured{}
+		sqlInstance.SetGroupVersionKind(krm.SQLInstanceGVK)
+		if err := kube.Get(ctx, key, sqlInstance); err != nil {
+			if apierrors.IsNotFound(err) {
+				return "", k8s.NewReferenceNotFoundError(krm.SQLInstanceGVK, key)
+			}
+			return "", fmt.Errorf("error reading referenced SQLInstance %v: %w", key, err)
+		}
+
+		sqlInstanceName, err := refs.GetResourceID(sqlInstance)
+		if err != nil {
+			return "", err
+		}
+		return sqlInstanceName, nil
+	} else {
+		return "", fmt.Errorf("must specify either spec.settings.cloneSource.sqlInstanceRef.external or spec.settings.cloneSource.sqlInstanceRef.name")
+	}
+}
+
+func normalizeSourceSQLDatabaseRefs(ctx context.Context, kube client.Reader, obj *krm.SQLInstance) ([]string, error) {
+	if obj.Spec.CloneSource == nil || obj.Spec.CloneSource.SQLDatabaseRefs == nil {
+		return nil, nil
+	}
+
+	sqlDatabaseRefs := obj.Spec.CloneSource.SQLDatabaseRefs
+	sqlDatabases := make([]string, len(sqlDatabaseRefs))
+
+	for i, sqlDatabaseRef := range sqlDatabaseRefs {
+		if sqlDatabaseRef.External != "" && sqlDatabaseRef.Name != "" {
+			return nil, fmt.Errorf("cannot specify both spec.settings.cloneSource.sqlDatabaseRefs[].external and spec.settings.cloneSource.sqlDatabaseRefs[].name")
+		}
+
+		if sqlDatabaseRef.External != "" {
+			sqlDatabases[i] = sqlDatabaseRef.External
```

Comment:
One gotcha here is that I expect (some) users will use the full path for most external databases (e.g. `projects/foo/locations/bar/instances/baz/databases/bat` ?

Reply:
Agreed it's probably simpler to just use a list of strings of database names. Fixed in https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/2429.